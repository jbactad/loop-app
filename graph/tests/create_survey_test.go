package tests_test

import (
	"testing"
	"time"

	"github.com/99designs/gqlgen/client"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/jbactad/loop/application/ports/mocks"
	"github.com/jbactad/loop/graph/models"
	"github.com/stretchr/testify/assert"
)

func TestCreateSurvey(t *testing.T) {
	type args struct {
		query string
		input models.NewSurvey
	}
	tests := []struct {
		name    string
		args    args
		wantErr assert.ErrorAssertionFunc
		setup   func(tp *mocks.TimeProvider, uig *mocks.UUIDGenerator)
	}{
		{
			name: "with valid query, it should return survey without error",
			args: args{
				query: `mutation CreateSurvey($input: NewSurvey!) {
  createSurvey(input: $input) {
    createdAt
    description
    id
    name
    question
    updatedAt
  }
}`,
				input: models.NewSurvey{
					Name:        "Survey 1",
					Description: "Survey 1 description",
					Question:    "Survey 1 question",
				},
			},
			setup: func(tp *mocks.TimeProvider, uig *mocks.UUIDGenerator) {
				tn := time.Date(2023, 1, 27, 12, 0, 0, 0, time.UTC)
				tp.EXPECT().Now().Return(tn)
				uid := "123e4567-e89b-12d3-a456-426614174000"
				uig.EXPECT().Generate().Return(uid)
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			timeProvider := mocks.NewTimeProvider(t)
			uuidGenerator := mocks.NewUUIDGenerator(t)

			c := NewTestClient(t, timeProvider, uuidGenerator)

			if tt.setup != nil {
				tt.setup(timeProvider, uuidGenerator)
			}

			var resp map[string]interface{}
			err := c.Post(tt.args.query, &resp, client.Var("input", tt.args.input))
			if tt.wantErr != nil {
				tt.wantErr(t, err)
				return
			}

			snaps.MatchJSON(t, resp)
		})
	}
}
