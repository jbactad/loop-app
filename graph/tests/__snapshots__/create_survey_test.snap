
[TestCreateSurvey/with_valid_query,_it_should_return_survey_without_error - 1]
{
 "createSurvey": {
  "createdAt": "2023-01-27T12:00:00Z",
  "description": "Survey 1 description",
  "id": "123e4567-e89b-12d3-a456-426614174000",
  "name": "Survey 1",
  "question": "Survey 1 question",
  "updatedAt": "2023-01-27T12:00:00Z"
 }
}
---
